### PR TITLE
Move read bytes utilities to Encoding

### LINKED ISSUE #521: Clean Parser API
The current `Parser` API consist of the expected `parse` methods and some `read...` methods.
I think the `read...` methods should not be public (at least not available from the `Parser` class).

### PR DESCRIPTION
**Pull request by andreaTP**
Fix #521

### CODE REVIEW

File: wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java [hunk 1]
```diff
@@ -11,6 +13,126 @@ public final class Encoding {
...
+    public static String readName(ByteBuffer buffer) {
+        return readName(buffer, true);
+    }
+
+    public static String readName(ByteBuffer buffer, boolean checkMalformed) {
+        var length = (int) readVarUInt32(buffer);
+        byte[] bytes = new byte[length];
+        try {
+            readBytes(buffer, bytes);
+        } catch (BufferUnderflowException e) {
```
Comment (electrum) [inline wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java:122]: This catch shouldn't be needed as `readBytes` already checks the length.

Or we could replace the checks in those `read` methods with handling for `BufferUnderflowException`.

File: wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java [hunk 2]
```diff
@@ -11,6 +13,126 @@ public final class Encoding {
...
+     * See <a href="https://www.w3.org/TR/wasm-core-1/#names%E2%91%A0">2.2.4. Names</a> of the WebAssembly Core Specification.
+     *
+     * @param buffer the byte buffer
+     * @return the resulting long
+     */
+    public static String readName(ByteBuffer buffer) {
+        return readName(buffer, true);
+    }
+
+    public static String readName(ByteBuffer buffer, boolean checkMalformed) {
```
Comment (electrum) [inline wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java:117]: Nit: copy Javadoc from above or make it `@see` the above

File: wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java [hunk 3]
```diff
@@ -11,6 +13,126 @@ public final class Encoding {
...
+     * @param buffer the byte buffer
+     * @return the resulting long
+     */
+    public static long readFloat32(ByteBuffer buffer) {
+        return readInt(buffer);
+    }
+
+    /**
+     * Read a symbol name from the buffer as UTF-8 String.
+     * See <a href="https://www.w3.org/TR/wasm-core-1/#names%E2%91%A0">2.2.4. Names</a> of the WebAssembly Core Specification.
```
Comment (electrum) [inline wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java:108]: The link is broken: the extra `%E2%91%A0` at the end of the URL shouldn't be there.

File: wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java [hunk 4]
```diff
@@ -11,6 +13,126 @@ public final class Encoding {
 
     private Encoding() {}
 
+    static int readInt(ByteBuffer buffer) {
```
Comment (electrum) [inline wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java:16]: Do these need to be public? Could they be needed for parsing custom sections?

This is just a question to consider. It's fine to leave them private for now.

**Review by electrum — APPROVED** [review]

File: wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java (hunk 1 above)
Comment (andreaTP) [reply wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java:122]: removed

File: wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java (hunk 2 above)
Comment (andreaTP) [reply wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java:117]: done

File: wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java (hunk 3 above)
Comment (andreaTP) [reply wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java:108]: good catch, fixed!

File: wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java (hunk 4 above)
Comment (andreaTP) [reply wasm/src/main/java/com/dylibso/chicory/wasm/Encoding.java:16]: In the current status I think that there are quite a few changes necessary to offer a proper, usable and generic API to parse custom sections ...
I would not try to predict the future, and welcome effort in this direction when someone needs it.

Opening visibility is not going to be a breaking change.

**Pull request merged**